(GH-2375) Bolt 3.0 developer update (#2389)
puppetlabs/bolt: documentation/developer_updates.md

## November 2020

### Changes coming in Bolt 3.0

With the new year fast approaching, the team is preparing to release the next

> major version of Bolt: **Bolt 3.0**! Over the next few releases, you will start

Reviewer: ```suggestion
major version of Bolt: **Bolt 3.0**! Over the next few releases you will start
```

to see deprecation warnings for specific features and configuration options
that are slated for removal in Bolt 3.0. To help you prepare for the release,
we've compiled a list of expected changes and removals.

#### Changes

- **Ship with Puppet 7**

  Bolt packages will ship with Puppet 7. This change should not impact most

>   users, but changes to the plan language in Puppet may require updating your

Reviewer: ```suggestion
  users, but changes to the Puppet language may require updating your Bolt
```

I think this is also technically true for yaml plans, since some changes will impact resource steps.

  plans.

  To read more about the changes in Puppet 7, see the [Puppet 7 release
  notes](https://puppet.com/docs/puppet/7.0/release_notes_puppet.html#release_notes_puppet_x-0-0).

- **Ship with Ruby 2.7**

  Bolt packages will ship with Ruby 2.7 instead of Ruby 2.5.1. This change will
  not affect most users who install the Bolt package, as Bolt packages ship with
  their own Ruby. However, gem installs will require Ruby 2.7 or higher, and
  users who install gems with Bolt's Ruby may also need to update their gems.

- **New default modulepath**

  Since the introduction of the module feature in Bolt 2.30.0, Bolt has used a
  shorter, simpler default modulepath when a project enables the feature.
  Starting in Bolt 3.0, the module management feature will be enabled for all

>   projects, meaning the default modulepath will also be updated.

Reviewer: It seems useful to add that users can update using `bolt project migrate`, either now or when 3.0 is released.

Author: Heh, I actually had that in here originally but pulled it out. :P I feel like this happens all the time.

  Puppetfile. The new name more clearly indicates that it configures how
  modules are installed.

  For a full list of project configuration options, see the [bolt-project.yaml
  options](bolt_project_reference.md) reference page.

#### Removals

- **Packages for Debian 8 will be removed**

  Debian 8 reached end-of-life on June 30, 2020.

  For a full list of supported platforms, see [installing Bolt](bolt_installing.md).

- **Support for `bolt.yaml` configuration file will be removed**

  Bolt will no longer support the `bolt.yaml` configuration file. Instead, you
  should use `bolt-project.yaml` in your project and `bolt-defaults.yaml` at
  the user and system level.

  For more information about configuring your project, see [configuring
  Bolt](configuring_bolt.md).

- **`bolt puppetfile *` commands and `*-BoltPuppetfile` PowerShell cmdlets will be removed**

  These commands are being removed in favor of the new `bolt module *` commands
  and `*-BoltModule` PowerShell cmdlets.

  To read more about the new module management workflow and the updated
  modulepath, see the [modules overview](modules.md).

> - **Support for running on remote targets with PowerShell 2.0 will be dropped**

Reviewer: And controllers, but I think that can be in the text.

  PowerShell 2.0 has been deprecated since 2017 and adds unecessary complexity
  for Bolt when connecting to remote targets running PowerShell 2.0.

- **`--boltdir` command-line option will be removed**

  This command-line option has been deprecated in favor of `--project`.

- **`--configfile` command-line option will be removed**

  This command-line option provides little value and adds unecessary complexity
  to Bolt. Instead, use the `--project` command-line option to run Bolt from a
  specific project.

- **`--debug` command-line option will be removed**

  This command-line option has been deprecated in favor of `--log-level debug`.
  Additionally, Bolt now writes a debug log to `bolt-debug.log` by default,
  removing the need to explicitly run in debug mode for most users.

- **`--description` command-line option will be removed**

  This command-line option provides little value.

> - **`target` key for YAML plan steps will be removed**

Reviewer: ```suggestion
- **Rename `target` key to `targets` for YAML plan steps**
```
This sounds less catastrophic.

  This key has been deprecated in favor of the `targets` key.

  For more information about writing YAML plans, see [writing YAML
  plans](writing_yaml_plans.md).

> - **`source` key for YAML plan `upload` step will be removed**

Reviewer: ```suggestion
- **Rename `source` key for YAML plan `upload` step to `upload`**
```

  This key has been deprecated in favor of the `upload` key.

  For more information about upload YAML plan step, see [writing YAML
  plans](writing_yaml_plans.md#file-upload-step).

> - **`private-key` and `public-key` parameters for pkcs7 plugin will be removed**

Reviewer: Same rename

- **`--description` command-line option will be removed**

  This command-line option provides little value.

- **`inventoryfile` configuration option will be removed**

  Bolt already offers the `--inventoryfile` command-line option to use a
  non-default inventory file for a single run. However, there is little value in
  configuring a permanent non-default inventoryfile for a project.

#### Updating your project

While some of these changes will require you to manually update files, others
can be made automatically with Bolt. We'll be updating the project migrate
command to automatically update your Bolt project to use the latest features and
best practices. You can run this command now to pick up any changes that have
already been made, or wait until Bolt 3.0 is released to pick up all changes at
once.

To migrate your project, run the following command in your project directory:

_\*nix shell command_

```shell
bolt project migrate
```

_PowerShell cmdlet_

```powershell
Update-BoltProject
```

Keep in mind that this command only makes changes to a project. Changes to plans
and configuration files at the user and system level will need to be made

> manually.

Reviewer 2: ```suggestion
manually. For more information on how the `migrate` command changes your project files, see [Migrate a Bolt project](projects.md#migrate-a-bolt-project).
```
